File: backend/agents/processor.py
"""
Agent message processing and orchestration
"""

from typing import Dict, Any, List, Optional
from uuid import UUID
import logging
from sqlalchemy.orm import Session

from core.database import SessionLocal
from models.database import Agent, Message
from agents.specialists import (
    BackendAgent,
    FrontendAgent,
    QAAgent,
    DevOpsAgent,
    OrchestratorAgent
)
from utils.text_parsing import extract_mentions

logger = logging.getLogger(__name__)

# Agent registry
AGENT_CLASSES = {
    "backend": BackendAgent,
    "frontend": FrontendAgent,
    "qa": QAAgent,
    "devops": DevOpsAgent,
    "orchestrator": OrchestratorAgent
}

# Cache for instantiated agents
_agent_cache: Dict[UUID, Any] = {}


def clear_agent_cache():
    """
    Clear the agent cache.
    This is called on application startup to ensure fresh agent instances
    with valid HTTP client connections after hot-reload.
    """
    global _agent_cache
    logger.info(f"Clearing agent cache ({len(_agent_cache)} agents)")
    _agent_cache.clear()


async def cleanup_agent_cache():
    """
    Cleanup all cached agents, closing their HTTP clients properly.
    This is called on application shutdown.
    """
    global _agent_cache
    logger.info(f"Cleaning up agent cache ({len(_agent_cache)} agents)")

    for agent_id, agent in _agent_cache.items():
        try:
            # Close httpx client if it exists (for Ollama)
            if hasattr(agent, 'llm') and hasattr(agent.llm, 'client'):
                if hasattr(agent.llm.client, 'aclose'):
                    await agent.llm.client.aclose()
                    logger.debug(f"Closed HTTP client for agent {agent.name}")
        except Exception as e:
            logger.warning(f"Error closing client for agent {agent_id}: {e}")

    _agent_cache.clear()


# ============================================
# Agent Task Queue Functions
# ============================================

def can_accept_task(agent_id: UUID, db: Session) -> bool:
    """
    Check if an agent can accept a new task (not currently busy)

    Args:
        agent_id: Agent UUID
        db: Database session

    Returns:
        True if agent is available, False if busy
    """
    agent = db.query(Agent).filter(Agent.id == agent_id).first()
    if not agent:
        return False

    return not agent.is_busy


def mark_agent_busy(agent_id: UUID, task_id: UUID, db: Session):
    """
    Mark an agent as busy with a specific task

    Args:
        agent_id: Agent UUID
        task_id: Task/Message UUID the agent is working on
        db: Database session
    """
    agent = db.query(Agent).filter(Agent.id == agent_id).first()
    if agent:
        agent.is_busy = True
        agent.current_task_id = task_id
        db.commit()
        logger.info(f"Agent {agent.name} marked as busy with task {task_id}")


def mark_agent_available(agent_id: UUID, db: Session):
    """
    Mark an agent as available (finished current task)

    Args:
        agent_id: Agent UUID
        db: Database session
    """
    agent = db.query(Agent).filter(Agent.id == agent_id).first()
    if agent:
        task_id = agent.current_task_id
        agent.is_busy = False
        agent.current_task_id = None
        db.commit()
        logger.info(f"Agent {agent.name} marked as available (completed task {task_id})")


def get_agent_instance(agent_record: Agent) -> Any:
    """
    Get or create agent instance from database record
    """
    if agent_record.id in _agent_cache:
        return _agent_cache[agent_record.id]

    agent_class = AGENT_CLASSES.get(agent_record.agent_type)
    if not agent_class:
        logger.error(f"Unknown agent type: {agent_record.agent_type}")
        return None

    # Create agent instance
    agent = agent_class(
        agent_id=agent_record.id,
        name=agent_record.name,
        agent_type=agent_record.agent_type,
        persona=agent_record.persona,
        config=agent_record.config
    )

    _agent_cache[agent_record.id] = agent
    return agent


async def process_agent_message(
    message_id: UUID,
    content: str,
    channel_id: UUID,
    mentioned_agents: List[str],
    manager: Any,  # WebSocket manager
    depth: int = 0,  # Recursion depth (max 3)
    call_chain: Optional[List[str]] = None  # Track agent call chain to prevent loops
):
    """
    Process a message that mentions one or more agents

    Args:
        message_id: UUID of the message that triggered this
        content: Message content
        channel_id: Channel UUID
        mentioned_agents: List of agent names (without @)
        manager: WebSocket manager for broadcasting
        depth: Current recursion depth (0 = user message, 1 = agent response, etc.)
        call_chain: List of agent names already called in this chain (loop prevention)
    """
    from models.database import Channel

    # Initialize call chain if not provided
    if call_chain is None:
        call_chain = []

    # Check recursion depth limit
    MAX_DEPTH = 3
    if depth >= MAX_DEPTH:
        logger.warning(f"Max recursion depth ({MAX_DEPTH}) reached, stopping agent chain")
        return

    db = SessionLocal()
    try:
        # Get channel to check context_cleared_at
        channel = db.query(Channel).filter(Channel.id == channel_id).first()

        # Build message query
        message_query = db.query(Message).filter(Message.channel_id == channel_id)

        # If context was cleared, only get messages after that timestamp
        if channel and channel.context_cleared_at:
            message_query = message_query.filter(Message.created_at > channel.context_cleared_at)

        # Get conversation context
        recent_messages = (
            message_query
            .order_by(Message.created_at.desc())
            .limit(10)
            .all()
        )

        context = {
            "conversation_history": [
                {
                    "author": msg.author_name,
                    "content": msg.content,
                    "type": msg.author_type
                }
                for msg in reversed(recent_messages)
            ],
            "depth": depth,
            "call_chain": call_chain
        }

        # Process each mentioned agent
        for agent_name in mentioned_agents:
            # Check if agent is already in call chain (prevent loops)
            if agent_name in call_chain:
                logger.warning(f"Agent @{agent_name} already in call chain {call_chain}, skipping to prevent loop")
                continue
            # Get agent from database
            agent_record = db.query(Agent).filter(
                Agent.name == f"@{agent_name}"
            ).first()

            if not agent_record or not agent_record.is_active:
                logger.warning(f"Agent not found or inactive: @{agent_name}")
                continue

            # Check if agent is busy
            if agent_record.is_busy:
                logger.warning(f"Agent @{agent_name} is busy with task {agent_record.current_task_id}, queueing not implemented yet")
                # TODO: Implement task queueing
                # For now, send a status message
                busy_message = Message(
                    channel_id=channel_id,
                    author_id=None,
                    author_type='system',
                    author_name='System',
                    content=f"⏳ @{agent_name} is currently busy. Please try again in a moment.",
                    msg_metadata={'busy': True, 'agent': agent_name}
                )
                db.add(busy_message)
                db.commit()
                db.refresh(busy_message)
                await manager.broadcast('message_new', {
                    'id': str(busy_message.id),
                    'channel_id': str(busy_message.channel_id),
                    'author_type': busy_message.author_type,
                    'author_name': busy_message.author_name,
                    'content': busy_message.content,
                    'created_at': busy_message.created_at.isoformat(),
                    'metadata': busy_message.msg_metadata
                })
                continue

            # Mark agent as busy
            mark_agent_busy(agent_record.id, message_id, db)

            # Send "thinking" status
            await manager.broadcast('agent_status', {
                'agent_name': f"@{agent_name}",
                'status': 'thinking'
            })

            try:
                # Special handling for orchestrator: Use workflow system for complex tasks
                if agent_record.agent_type == "orchestrator":
                    # Import here to avoid circular dependency
                    from agents.workflow_orchestrator import WorkflowOrchestrator

                    logger.info(f"Orchestrator triggered via chat, creating workflow for: {content[:100]}")

                    # Create workflow orchestrator
                    workflow_orchestrator = WorkflowOrchestrator(manager)

                    # Create workflow from user request
                    workflow = await workflow_orchestrator.create_workflow_from_request(
                        user_request=content,
                        orchestrator_id=agent_record.id,
                        channel_id=channel_id,
                        db=db
                    )

                    # Post workflow created message to channel
                    workflow_msg = Message(
                        channel_id=channel_id,
                        author_id=agent_record.id,
                        author_type='agent',
                        author_name=agent_record.name,
                        content=f"I've created a workflow to handle your request. Executing {len(workflow.workflow_tasks)} tasks...",
                        metadata={
                            'workflow_id': str(workflow.id),
                            'in_reply_to': str(message_id)
                        }
                    )
                    db.add(workflow_msg)
                    db.commit()
                    db.refresh(workflow_msg)

                    await manager.broadcast('message_new', {
                        'id': str(workflow_msg.id),
                        'channel_id': str(workflow_msg.channel_id),
                        'author_type': workflow_msg.author_type,
                        'author_name': workflow_msg.author_name,
                        'content': workflow_msg.content,
                        'created_at': workflow_msg.created_at.isoformat(),
                        'metadata': workflow_msg.msg_metadata
                    })

                    # Execute workflow (async in background with new DB session)
                    import asyncio

                    async def run_workflow():
                        """Execute workflow with its own database session"""
                        workflow_db = SessionLocal()
                        try:
                            await workflow_orchestrator.execute_workflow(workflow.id, workflow_db)
                        finally:
                            workflow_db.close()

                    asyncio.create_task(run_workflow())

                    # Mark orchestrator as available (workflow runs in background)
                    mark_agent_available(agent_record.id, db)

                    await manager.broadcast('agent_status', {
                        'agent_name': f"@{agent_name}",
                        'status': 'online'
                    })

                    # Skip regular message processing for orchestrator
                    continue

                # Regular agent processing for non-orchestrator agents
                # Get agent instance
                agent = get_agent_instance(agent_record)
                if not agent:
                    mark_agent_available(agent_record.id, db)
                    continue

                # Process message
                response = await agent.process_message(content, context)

                # Save agent's response to database
                agent_message = Message(
                    channel_id=channel_id,
                    author_id=agent_record.id,
                    author_type='agent',
                    author_name=agent_record.name,
                    content=response,
                    metadata={
                        'model': agent.llm.model,
                        'provider': agent.llm.provider,
                        'in_reply_to': str(message_id)
                    }
                )
                db.add(agent_message)
                db.commit()
                db.refresh(agent_message)

                # Broadcast agent response
                await manager.broadcast('message_new', {
                    'id': str(agent_message.id),
                    'channel_id': str(agent_message.channel_id),
                    'author_type': agent_message.author_type,
                    'author_name': agent_message.author_name,
                    'content': agent_message.content,
                    'created_at': agent_message.created_at.isoformat(),
                    'metadata': agent_message.msg_metadata
                })

                # Mark agent as available
                mark_agent_available(agent_record.id, db)

                # Update agent status back to online
                await manager.broadcast('agent_status', {
                    'agent_name': f"@{agent_name}",
                    'status': 'online'
                })

                # Check agent response for @mentions (recursive triggering)
                # Extract mentions from response
                response_mentions = extract_mentions(response, strip_md=True)

                if response_mentions and depth < MAX_DEPTH - 1:
                    # Found mentions in agent response - trigger those agents recursively
                    sub_agent_names = [mention[1] for mention in response_mentions]  # Extract names without @

                    # Filter out mentions already in call chain
                    new_agents = [name for name in sub_agent_names if name not in call_chain and name != agent_name]

                    if new_agents:
                        logger.info(f"Agent @{agent_name} mentioned {new_agents}, triggering recursively (depth {depth + 1})")

                        # Update call chain for sub-agents
                        updated_chain = call_chain + [agent_name]

                        # Recursively process sub-agent mentions
                        import asyncio
                        asyncio.create_task(
                            process_agent_message(
                                message_id=agent_message.id,  # Use agent's message as trigger
                                content=response,
                                channel_id=channel_id,
                                mentioned_agents=new_agents,
                                manager=manager,
                                depth=depth + 1,
                                call_chain=updated_chain
                            )
                        )

            except Exception as e:
                logger.error(f"Error processing agent {agent_name}: {e}")

                # Mark agent as available (release from busy state)
                mark_agent_available(agent_record.id, db)

                # Send error message
                error_message = Message(
                    channel_id=channel_id,
                    author_id=agent_record.id,
                    author_type='agent',
                    author_name=agent_record.name,
                    content=f"I encountered an error: {str(e)}",
                    msg_metadata={'error': str(e), 'in_reply_to': str(message_id)}
                )
                db.add(error_message)
                db.commit()
                db.refresh(error_message)

                await manager.broadcast('message_new', {
                    'id': str(error_message.id),
                    'channel_id': str(error_message.channel_id),
                    'author_type': error_message.author_type,
                    'author_name': error_message.author_name,
                    'content': error_message.content,
                    'created_at': error_message.created_at.isoformat(),
                    'metadata': error_message.msg_metadata
                })

                # Update status to online
                await manager.broadcast('agent_status', {
                    'agent_name': f"@{agent_name}",
                    'status': 'online'
                })

    finally:
        db.close()


async def invoke_agent(
    agent_name: str,
    message: str,
    context: Dict[str, Any] = None,
    channel_id: Optional[UUID] = None
) -> Dict[str, Any]:
    """
    Directly invoke an agent by name
    """
    db = SessionLocal()
    try:
        # Add @ prefix if not present
        if not agent_name.startswith('@'):
            agent_name = f'@{agent_name}'

        # Get agent from database
        agent_record = db.query(Agent).filter(Agent.name == agent_name).first()
        if not agent_record:
            raise ValueError(f"Agent not found: {agent_name}")

        if not agent_record.is_active:
            raise ValueError(f"Agent is not active: {agent_name}")

        # Get agent instance
        agent = get_agent_instance(agent_record)
        if not agent:
            raise ValueError(f"Could not instantiate agent: {agent_name}")

        # Process message
        response = await agent.process_message(message, context or {})

        # Save to database if channel provided
        if channel_id:
            agent_message = Message(
                channel_id=channel_id,
                author_id=agent_record.id,
                author_type='agent',
                author_name=agent_record.name,
                content=response,
                metadata={
                    'model': agent.llm.model,
                    'provider': agent.llm.provider
                }
            )
            db.add(agent_message)
            db.commit()
            db.refresh(agent_message)

            return {
                'id': str(agent_message.id),
                'channel_id': str(agent_message.channel_id),
                'author_type': agent_message.author_type,
                'author_name': agent_message.author_name,
                'content': agent_message.content,
                'created_at': agent_message.created_at.isoformat(),
                'metadata': agent_message.msg_metadata
            }

        return {
            'agent_name': agent_name,
            'response': response,
            'metadata': {
                'model': agent.llm.model,
                'provider': agent.llm.provider
            }
        }

    finally:
        db.close()


async def invoke_agent_by_id(
    agent_id: UUID,
    message: str,
    context: Dict[str, Any] = None,
    channel_id: Optional[UUID] = None
) -> Dict[str, Any]:
    """
    Directly invoke an agent by ID
    """
    db = SessionLocal()
    try:
        # Get agent from database
        agent_record = db.query(Agent).filter(Agent.id == agent_id).first()
        if not agent_record:
            raise ValueError(f"Agent not found with ID: {agent_id}")

        return await invoke_agent(agent_record.name, message, context, channel_id)

    finally:
        db.close()

File: backend-ts/src/agents/processor.ts
/**
 * Message Processing Pipeline — @mention routing, agent invocation,
 * streaming relay, and recursive delegation.
 *
 * Replaces backend/agents/processor.py (686 LOC).
 *
 * Pipeline flow:
 *   1. User sends message (via WebSocket)
 *   2. Extract @mentions from message content
 *   3. For each mentioned agent:
 *      a. Look up agent record in database
 *      b. Check availability (is_active, !is_busy)
 *      c. Mark agent busy + broadcast 'agent_status' thinking
 *      d. Build context from recent messages
 *      e. Process message (streaming or non-streaming)
 *      f. Save agent response to database
 *      g. Broadcast 'message_new' / 'message_stream' / 'message_update'
 *      h. Check agent response for @mentions (recursive delegation)
 *      i. Mark agent available + broadcast 'agent_status' online
 */

import { randomUUID } from "crypto";
import { eq, desc, and, gt } from "drizzle-orm";
import { db } from "../db/connection.js";
import { agents, messages, channels } from "../db/schema.js";
import { parseMentions, type AgentName } from "./tools/delegation.js";
import {
  RezNetAgent,
  type ProcessMessageContext,
  type AgentPersona,
} from "./base-agent.js";
import type { AgentEvent } from "@mariozechner/pi-agent-core";
import { wsManager } from "../websocket/manager.js";
import { resolveModel } from "../llm/client.js";
import { createSpecialistAgent } from "./specialists.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Maximum recursion depth for agent-to-agent delegation. */
const MAX_DELEGATION_DEPTH = 3;

// ---------------------------------------------------------------------------
// Agent cache
// ---------------------------------------------------------------------------

const agentCache = new Map<string, RezNetAgent>();

/**
 * Get or create an agent instance (cached for reuse).
 */
function getOrCreateAgent(
  agentId: string,
  agentType: string,
  _persona: unknown,
): RezNetAgent {
  if (agentCache.has(agentId)) {
    return agentCache.get(agentId)!;
  }

  // Determine if this is a known specialist type
  const knownType = agentType as AgentName;
  const isKnown = ["orchestrator", "backend", "frontend", "qa", "devops"].includes(knownType);

  let agent: RezNetAgent;

  if (isKnown) {
    agent = createSpecialistAgent({
      agentId,
      agentType: knownType,
      model: resolveModel(),
      onDelegate: async (target, task) => {
        // Recursive delegation via agent cache
        const targetAgent = agentCache.get(`agent-${target}`);
        if (targetAgent) {
          return targetAgent.processMessage(task);
        }
        return `Agent @${target} is not available.`;
      },
    });
  } else {
    // Custom agent — use persona from database
    const persona = (_persona as AgentPersona) ?? {
      role: `a ${agentType} specialist`,
      goal: "Help with tasks related to your expertise.",
      backstory: "You are a specialist agent.",
      capabilities: ["General assistance"],
      color: "#888888",
    };

    agent = new RezNetAgent({
      agentId,
      name: agentType,
      agentType,
      persona,
      model: resolveModel(),
    });
  }

  agentCache.set(agentId, agent);
  return agent;
}

/**
 * Clear the agent cache (e.g., on shutdown).
 */
export function clearAgentCache(): void {
  agentCache.clear();
}

// ---------------------------------------------------------------------------
// Context builder
// ---------------------------------------------------------------------------

/**
 * Build conversation context from recent messages in a channel.
 */
async function buildMessageContext(
  channelId: string,
  _depth: number,
  _callChain: string[],
): Promise<ProcessMessageContext> {
  // Get channel to check context_cleared_at
  const channelRows = await db
    .select()
    .from(channels)
    .where(eq(channels.id, channelId))
    .limit(1);
  const channel = channelRows[0];

  // Build base query conditions
  const conditions = [eq(messages.channelId, channelId)];

  // Only include messages after context was cleared
  if (channel?.contextClearedAt) {
    conditions.push(gt(messages.createdAt, channel.contextClearedAt));
  }

  // Get recent 10 messages
  const recentMessages = await db
    .select()
    .from(messages)
    .where(and(...conditions))
    .orderBy(desc(messages.createdAt))
    .limit(10);

  // Reverse to chronological order
  const conversationHistory = recentMessages.reverse().map((msg) => ({
    role: msg.authorType,
    content: msg.content,
    author: msg.authorName ?? undefined,
  }));

  return {
    conversationHistory,
    channelId,
  };
}

// ---------------------------------------------------------------------------
// Message serializer
// ---------------------------------------------------------------------------

/**
 * Serialize a message record for WebSocket broadcast.
 */
function serializeMessage(msg: {
  id: string;
  channelId: string;
  authorType: string;
  authorName: string | null;
  authorId: string | null;
  content: string;
  createdAt: Date | null;
  threadId: string | null;
  msgMetadata: unknown;
}): Record<string, unknown> {
  return {
    id: msg.id,
    channel_id: msg.channelId,
    author_type: msg.authorType,
    author_name: msg.authorName,
    author_id: msg.authorId,
    content: msg.content,
    created_at: msg.createdAt?.toISOString() ?? new Date().toISOString(),
    thread_id: msg.threadId,
    metadata: msg.msgMetadata ?? {},
  };
}

// ---------------------------------------------------------------------------
// Core: processAgentMessage
// ---------------------------------------------------------------------------

export interface ProcessAgentMessageOptions {
  messageId: string;
  content: string;
  channelId: string;
  mentionedAgents: AgentName[];
  depth?: number;
  callChain?: string[];
}

/**
 * Process a message that mentions one or more agents.
 *
 * For each mentioned agent:
 *   1. Look up in database
 *   2. Check availability
 *   3. Process message with streaming
 *   4. Save response
 *   5. Check for recursive delegation
 */
export async function processAgentMessage(
  opts: ProcessAgentMessageOptions,
): Promise<void> {
  const {
    messageId,
    content,
    channelId,
    mentionedAgents,
    depth = 0,
    callChain = [],
  } = opts;

  // Depth guard
  if (depth >= MAX_DELEGATION_DEPTH) {
    console.warn(
      `[processor] Max recursion depth (${MAX_DELEGATION_DEPTH}) reached, stopping agent chain`,
    );
    return;
  }

  // Process each mentioned agent
  for (const agentName of mentionedAgents) {
    try {
      await processOneAgent({
        agentName,
        messageId,
        content,
        channelId,
        depth,
        callChain,
      });
    } catch (err) {
      console.error(
        `[processor] Error processing @${agentName}:`,
        err instanceof Error ? err.message : err,
      );
    }
  }
}

// ---------------------------------------------------------------------------
// Process a single agent
// ---------------------------------------------------------------------------

interface ProcessOneAgentOptions {
  agentName: AgentName;
  messageId: string;
  content: string;
  channelId: string;
  depth: number;
  callChain: string[];
}

async function processOneAgent(opts: ProcessOneAgentOptions): Promise<void> {
  const { agentName, messageId, content, channelId, depth, callChain } = opts;

  // 1. Look up agent in database
  const agentRows = await db
    .select()
    .from(agents)
    .where(eq(agents.name, agentName))
    .limit(1);

  const agentRecord = agentRows[0];
  if (!agentRecord) {
    console.warn(`[processor] Agent @${agentName} not found in database`);
    return;
  }

  // 2. Check availability
  if (!agentRecord.isActive) {
    console.warn(`[processor] Agent @${agentName} is inactive`);
    return;
  }

  if (agentRecord.isBusy) {
    // Send busy message
    const busyMsg = await db
      .insert(messages)
      .values({
        id: randomUUID(),
        channelId,
        authorType: "system",
        authorName: "System",
        content: `@${agentName} is currently busy. Please try again in a moment.`,
        msgMetadata: { busy: true, agent: agentName },
      })
      .returning();

    if (busyMsg[0]) {
      wsManager.broadcast("message_new", serializeMessage(busyMsg[0]));
    }
    return;
  }

  // 3. Mark agent as busy
  await db
    .update(agents)
    .set({ isBusy: true, currentTaskId: messageId })
    .where(eq(agents.id, agentRecord.id));

  // Broadcast thinking status
  wsManager.broadcast(
    "agent_status",
    { agent_name: `@${agentName}`, status: "thinking" },
    { batch: true },
  );

  try {
    // 4. Build context
    const context = await buildMessageContext(channelId, depth, callChain);

    // 5. Get or create agent instance
    const agent = getOrCreateAgent(
      agentRecord.id,
      agentRecord.name,
      agentRecord.persona,
    );

    // 6. Create placeholder message for streaming
    const placeholderMsg = await db
      .insert(messages)
      .values({
        id: randomUUID(),
        channelId,
        authorId: agentRecord.id,
        authorType: "agent",
        authorName: `@${agentRecord.name}`,
        content: "",
        msgMetadata: {
          in_reply_to: messageId,
          streaming: true,
        },
      })
      .returning();

    const placeholder = placeholderMsg[0];
    if (!placeholder) {
      throw new Error("Failed to create placeholder message");
    }

    // Broadcast empty placeholder
    wsManager.broadcast("message_new", serializeMessage(placeholder));

    // 7. Process with streaming
    let accumulatedResponse = "";

    const response = await agent.processMessageStreaming(
      content,
      context,
      (event: AgentEvent) => {
        if (event.type === "message_update") {
          // Extract text chunks from streaming
          const msg = event.message;
          if (msg && "content" in msg && Array.isArray(msg.content)) {
            for (const block of msg.content) {
              if (typeof block === "object" && "type" in block && block.type === "text" && "text" in block) {
                const text = block.text as string;
                if (text.length > accumulatedResponse.length) {
                  const chunk = text.slice(accumulatedResponse.length);
                  accumulatedResponse = text;

                  // Relay streaming chunk
                  wsManager.broadcast("message_stream", {
                    message_id: placeholder.id,
                    chunk,
                    is_final: false,
                  });
                }
              }
            }
          }
        }
      },
    );

    // Use final response
    accumulatedResponse = response;

    // 8. Update placeholder with complete content
    await db
      .update(messages)
      .set({
        content: accumulatedResponse,
        msgMetadata: {
          in_reply_to: messageId,
          streaming: false,
        },
      })
      .where(eq(messages.id, placeholder.id));

    // Broadcast final message update
    wsManager.broadcast("message_update", {
      id: placeholder.id,
      channel_id: channelId,
      author_type: "agent",
      author_name: `@${agentRecord.name}`,
      author_id: agentRecord.id,
      content: accumulatedResponse,
      created_at: placeholder.createdAt?.toISOString(),
      metadata: {
        in_reply_to: messageId,
        streaming: false,
      },
    });

    // Send final stream marker
    wsManager.broadcast("message_stream", {
      message_id: placeholder.id,
      chunk: "",
      is_final: true,
    });

    // 9. Check for recursive delegation
    if (depth < MAX_DELEGATION_DEPTH - 1) {
      const responseMentions = parseMentions(accumulatedResponse);
      const newAgents = responseMentions.filter(
        (name) => !callChain.includes(name) && name !== agentName,
      );

      if (newAgents.length > 0) {
        // Fire and forget — non-blocking recursive delegation
        void processAgentMessage({
          messageId: placeholder.id,
          content: accumulatedResponse,
          channelId,
          mentionedAgents: newAgents,
          depth: depth + 1,
          callChain: [...callChain, agentName],
        });
      }
    }
  } catch (err) {
    console.error(
      `[processor] Error in agent @${agentName}:`,
      err instanceof Error ? err.message : err,
    );

    // Save error message
    const errorContent =
      err instanceof Error
        ? `I encountered an error: ${err.message}`
        : "I encountered an unexpected error. Please try again.";

    const errorMsg = await db
      .insert(messages)
      .values({
        id: randomUUID(),
        channelId,
        authorId: agentRecord.id,
        authorType: "agent",
        authorName: `@${agentRecord.name}`,
        content: errorContent,
        msgMetadata: {
          error: true,
          in_reply_to: messageId,
        },
      })
      .returning();

    if (errorMsg[0]) {
      wsManager.broadcast("message_new", serializeMessage(errorMsg[0]));
    }
  } finally {
    // 10. Mark agent available
    await db
      .update(agents)
      .set({ isBusy: false, currentTaskId: null })
      .where(eq(agents.id, agentRecord.id));

    wsManager.broadcast(
      "agent_status",
      { agent_name: `@${agentName}`, status: "online" },
      { batch: true },
    );
  }
}

// ---------------------------------------------------------------------------
// Direct agent invocation (bypasses @mention parsing)
// ---------------------------------------------------------------------------

export interface InvokeAgentOptions {
  agentName: string;
  message: string;
  channelId?: string;
  context?: Record<string, unknown>;
}

/**
 * Directly invoke a specific agent by name.
 * Used by the agent_invoke WebSocket event.
 */
export async function invokeAgent(
  opts: InvokeAgentOptions,
): Promise<{ agentName: string; response: string }> {
  const name = opts.agentName.startsWith("@")
    ? opts.agentName.slice(1)
    : opts.agentName;

  // If channel provided, route through full pipeline
  if (opts.channelId) {
    await processAgentMessage({
      messageId: randomUUID(),
      content: opts.message,
      channelId: opts.channelId,
      mentionedAgents: [name as AgentName],
    });
    return { agentName: name, response: "(processing via WebSocket)" };
  }

  // Direct invocation without channel — simpler path
  const agentRows = await db
    .select()
    .from(agents)
    .where(eq(agents.name, name))
    .limit(1);

  const agentRecord = agentRows[0];
  if (!agentRecord || !agentRecord.isActive) {
    throw new Error(`Agent @${name} not found or inactive`);
  }

  const agent = getOrCreateAgent(
    agentRecord.id,
    agentRecord.name,
    agentRecord.persona,
  );

  const response = await agent.processMessage(opts.message);
  return { agentName: name, response };
}
